velox/functions/sparksql/String.h

```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
   }
 };
 
+/// translate(string, matchStr, replaceStr) -> varchar
+///
+///   Returns a new translated string. It translates the character in ``string``
+///   by a character in ``replaceStr``. The character in ``replaceStr`` is
+///   corresponding to the character in ``matchStr``. The translation will
+///   happen when any character in ``string`` matching with the character in the
+///   ``matchStr``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  using MapType = std::unordered_map<StringView, StringView>;
+  MapType dict_;
+  bool isConstantTranslateDict = false;
+
+  FOLLY_ALWAYS_INLINE MapType buildDict(
+      const arg_type<Varchar>& matchStr,
+      const arg_type<Varchar>& replaceStr) {
+    MapType dict;
+    int i = 0, j = 0;
+    while (i < matchStr.size()) {
+      StringView replacePiece;
+      if (j >= replaceStr.size()) {
+        replacePiece = StringView("");
+      } else {
+        int replaceCharCount = utf8proc_char_length(replaceStr.data() + j);
+        replacePiece = StringView(replaceStr.data() + j, replaceCharCount);
+        j = j + replaceCharCount;
+      }
+      int matchCharCount = utf8proc_char_length(matchStr.data() + i);
+      StringView matchPiece = StringView(matchStr.data() + i, matchCharCount);
+      dict.try_emplace(matchPiece, replacePiece);
+      i = i + matchCharCount;
+    }
+    return dict;
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* matchStr,
+      const arg_type<Varchar>* replaceStr) {
+    if (matchStr != nullptr && replaceStr != nullptr) {
+      dict_ = buildDict(*matchStr, *replaceStr);
+      isConstantTranslateDict = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& matchStr,
+      const arg_type<Varchar>& replaceStr) {
+    if (!isConstantTranslateDict) {
+      dict_ = buildDict(matchStr, replaceStr);
+    }
+    // No need to do the replacement.
+    if (dict_.empty()) {
+      result.setNoCopy(StringView(input.data(), input.size()));
+      return;
+    }
+    int k = 0;
+    while (k < input.size()) {
```

Comment:
It looks like it would be helpful to add a function that loops over characters of a string and applies a lambda to each character. Or, create an iterator that returns next character.

Reply:
Yes, we can move this part into a function with utf-8/ ascii considered. Please take a review to see whether I fully caught your point. Thanks!